carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java

```diff
@@ -0,0 +1,489 @@
+import io.netty.channel.ChannelOption;
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.timeout.ReadTimeoutHandler;
+import io.netty.handler.timeout.WriteTimeoutHandler;
+import io.prometheus.client.Counter;
+import io.prometheus.client.Gauge;
+import java.net.InetSocketAddress;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import org.carapaceproxy.EndpointStats;
+import org.carapaceproxy.SimpleHTTPResponse;
+import org.carapaceproxy.server.mapper.MapResult;
+import org.carapaceproxy.client.EndpointKey;
+import org.carapaceproxy.server.cache.ContentsCache;
+import org.carapaceproxy.server.mapper.CustomHeader;
+import org.carapaceproxy.utils.PrometheusUtils;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.http.client.HttpClientResponse;
+import reactor.netty.resources.ConnectionProvider;
+
+/**
+ * Manager forwarding {@link ProxyRequest} from clients to proper endpoints.
+ *
+ * @author paolo.venturi
+ */
+public class ProxyRequestsManager {
+
+    public static final Counter USER_REQUESTS_COUNTER = PrometheusUtils.createCounter(
+            "listeners", "user_requests_total", "inbound requests count", "userId"
+    ).register();
+
+    public static final Gauge PENDING_REQUESTS_GAUGE = PrometheusUtils.createGauge(
+            "backends", "pending_requests", "pending requests"
+    ).register();
+
+    public static final Counter TOTAL_REQUESTS_COUNTER = PrometheusUtils.createCounter(
+            "backends", "sent_requests_total", "sent requests", "host"
+    ).register();
+
+    public static final Counter STUCK_REQUESTS_COUNTER = PrometheusUtils.createCounter(
+            "backends", "stuck_requests_total", "stuck requests, this requests will be killed"
+    ).register();
+
+    private static final Logger LOGGER = Logger.getLogger(ProxyRequestsManager.class.getName());
+
+    private final HttpProxyServer parent;
+    private ConnectionProvider connectionProvider;
+    private final ConcurrentHashMap<EndpointKey, EndpointStats> endpointsStats = new ConcurrentHashMap<>();
+
+    public ProxyRequestsManager(HttpProxyServer parent) {
+        this.parent = parent;
+    }
+
+    public ConcurrentHashMap<EndpointKey, EndpointStats> getEndpointsStats() {
```

Comment:
I wouldn't expose externally the map implementation, can it even be modified externally?